Move the CMS-utils and Navigation from apps to the common package

nikomakela opened the pull request:
Move the common headless CMS -utils and the Navigation-component from apps to the components -package.

The Navigation component was a good example how a simple looking common component needs huge refactoring when it is affected by the app specific routing and app-config. When the Navigation-component was moved to the common components-package, (the app specific) routing utilities e.g. `getI18nPath` and the CMS-utilities, needed to be moved in the common package as well. This led to a lot of refactoring, because:
1. The routing-utilities were app-scpecific
2. The cms-utilities were app-specific

It was needed that the utilities were configurable by the application, so they are now helper-classes, with all the old functions. The helper-class can be configured inside the application. So, every routingUtil import needed to be changed to routingHelper-instance-calls, etc.

The result is that now we need only a small app specific file that configures the routing-helper instance and same for cms-helper-instance. The new helper-utility-instances are provided with a new `CmsHelperProvider`.

----

The idea might be easies to understand when you start by looking at the changes done in the `_app.tsx` files

terovirtanen commented:
**Hobbies-Helsinki branch is deployed to platta: https://harrastukset-pr184.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Sports-Helsinki branch is deployed to platta: https://liikunta-pr184.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
## TestCafe result is __success__ for https://liikunta-pr184.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
## TestCafe result is __success__ for https://harrastukset-pr184.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
**Hobbies-Helsinki branch is deployed to platta: https://harrastukset-pr184.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
## TestCafe result is __success__ for https://harrastukset-pr184.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
**Hobbies-Helsinki branch is deployed to platta: https://harrastukset-pr184.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
## TestCafe result is __success__ for https://harrastukset-pr184.dev.hel.ninja :laughing::tada::tada::tada:

nenonja commented on apps/events-helsinki/tsconfig.json:
```diff
@@ -33,6 +33,7 @@
     "**/*.jsx",
-    "**/*.json"
+    "**/*.json",
+    "../hobbies-helsinki/src/domain/app/headlessCmsHelper.tsx"
```
what is this?

nikomakela replied:
some reason why the vscode adds to automatically. It's an accident. Nice catch!

nenonja commented on apps/hobbies-helsinki/src/pages/articles/[...slug].tsx:
```diff
@@ -88,9 +80,11 @@ const NextCmsArticle: NextPage<{
 
 export async function getStaticPaths() {
-  const articlePageInfos = await getAllArticles();
+  // NOTE: It might not be a good thing to use ApolloClient here,
```
A worth of a new ticket?

nenonja commented on apps/hobbies-helsinki/tsconfig.json:
```diff
@@ -33,6 +33,8 @@
     "**/*.jsx",
-    "**/*.json"
+    "**/*.json",
+    "src/domain/app/headlessCmsHelper.tsx",
```
why?

nikomakela replied:
some reason why the vscode adds to automatically. It's an accident. Nice catch!

nenonja commented on apps/sports-helsinki/src/pages/venues/[venueId]/index.tsx:
```diff
@@ -7,26 +11,25 @@ import type {
-import MatomoWrapper from '../../../domain/matomoWrapper/MatomoWrapper';
-import VenuePageContainer from '../../../domain/venue/VenuePageContainer';
-import { getLocaleOrError } from '../../../utils/routerUtils';
+import AppConfig from 'domain/app/AppConfig';
```
how thse non-relative paths can work?

nikomakela replied:
I have no idea 😄 It seems like the domain-path as an absolute path always works. Some constants and types does not, because it has multiple alternatives. Do we want to use these, or should I go through all those thousands of them? 😄 These were added by the Vscode and I just went from there where the fence was the lowest

nikomakela replied:
https://beqode.com/blog/absolute-vs-relative-import-paths-nodejs

nikomakela replied:
absolute path == relative path from the root

nenonja commented on packages/components/src/utils/headless-cms/HeadlessCMSHelper.tsx:
```diff
@@ -0,0 +1,239 @@
+
+  // NOTE: Referenced 2 times in next-pages
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  getUriID = (slugs: string[], locale: AppLanguage): string => {
```
I would rather use class functions inside a class instad of arrow functions, unless you plan to destruct the function from the class and pass the reference to it?
`getUriID(slugs: string[], locale: AppLanguage): string { ... }`

nenonja replied:
https://medium.com/@charpeni/arrow-functions-in-class-properties-might-not-be-as-great-as-we-think-3b3551c440b1

nenonja reviewed (commented):
Nice job and very welcomed refactoring! 
Some comments though, especially I wonder about the tsconfig changes.

nenonja commented:
Unit tests fail: https://dev.azure.com/City-of-Helsinki/harrastukset/_build/results?buildId=65042&view=logs&j=8267e077-3a31-5bba-93c5-4b3e55896ad0&t=f69267ae-8804-5590-a878-6662f5a5eb23

kristianjokela commented:
Just a general thought; now since there's provider & context for the helpers, would it make sense to just include all the functionality that's now inside `/domain/app` per app, in the common component context, and then pass the necessary dependencies like `AppConfig` etc. to the provider; then there wouldn't be need for `domain/app` and you could always call the helper functions through `useContext` or with hooks (which already do return them from context - so have similar useRouteHelper than useCmsHelper)? 

I was just thinking, since there is already provider, context & (some) hooks, why not just use them everywhere instead of importing those classes?

nikomakela committed:
refactor: move common router and CMS utils to components-package

HH-283
- add app config context for the common navigation component
- move router utils to from apps to components-package
- move the Headless CMS -utils from apps to components-package

nikomakela commented:
> Just a general thought; now since there's provider & context for the helpers, would it make sense to just include all the functionality that's now inside `/domain/app` per app, in the common component context, and then pass the necessary dependencies like `AppConfig` etc. to the provider; then there wouldn't be need for `domain/app` and you could always call the helper functions through `useContext` or with hooks (which already do return them from context - so have similar useRouteHelper than useCmsHelper)?
> 
> I was just thinking, since there is already provider, context & (some) hooks, why not just use them everywhere instead of importing those classes?

I'm not sure if I understood your idea. Did you mean that we would refactor the AppConfig to be provided from the context? Doesn't it mean that it would require a lot more refactoring than this? Could it be like a next step for this, or something?

I actually tried to make an interface for the AppConfig that could have been used with the Helper-classes, but JS/TS cannot have an interface-field for static methods, so I needed to drop that idea.

terovirtanen commented:
**Hobbies-Helsinki branch is deployed to platta: https://harrastukset-pr184.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Events-Helsinki branch is deployed to platta: https://tapahtumat-pr184.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Sports-Helsinki branch is deployed to platta: https://liikunta-pr184.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
## TestCafe result is __success__ for https://harrastukset-pr184.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
## TestCafe result is __success__ for https://tapahtumat-pr184.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
**$(upper(replace(parameters.proxy,'-', ' '))) is deployed to platta: https://events-graphql-proxy-pr184.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
## TestCafe result is __success__ for https://liikunta-pr184.dev.hel.ninja :laughing::tada::tada::tada:

kristianjokela commented:
> > Just a general thought; now since there's provider & context for the helpers, would it make sense to just include all the functionality that's now inside `/domain/app` per app, in the common component context, and then pass the necessary dependencies like `AppConfig` etc. to the provider; then there wouldn't be need for `domain/app` and you could always call the helper functions through `useContext` or with hooks (which already do return them from context - so have similar useRouteHelper than useCmsHelper)?
> > I was just thinking, since there is already provider, context & (some) hooks, why not just use them everywhere instead of importing those classes?
> 
> I'm not sure if I understood your idea. Did you mean that we would refactor the AppConfig to be provided from the context? Doesn't it mean that it would require a lot more refactoring than this? Could it be like a next step for this, or something?
> 

Yeah, I meant as next step - so we wouldn't need those "extra" helpers either, but...

> I actually tried to make an interface for the AppConfig that could have been used with the Helper-classes, but JS/TS cannot have an interface-field for static methods, so I needed to drop that idea.

..this might be a problem, I didn't realise those are static.. :)

But so I'm happy with how this is now, but was just wondering if it would be possible to use all those helpers through the context (but as next step).

terovirtanen commented:
**$(upper(replace(parameters.proxy,'-', ' '))) is deployed to platta: https://venue-graphql-proxy-pr184.dev.hel.ninja** :rocket::rocket::rocket:

nenonja commented:
> > Just a general thought; now since there's provider & context for the helpers, would it make sense to just include all the functionality that's now inside `/domain/app` per app, in the common component context, and then pass the necessary dependencies like `AppConfig` etc. to the provider; then there wouldn't be need for `domain/app` and you could always call the helper functions through `useContext` or with hooks (which already do return them from context - so have similar useRouteHelper than useCmsHelper)?
> > I was just thinking, since there is already provider, context & (some) hooks, why not just use them everywhere instead of importing those classes?
> 
> I'm not sure if I understood your idea. Did you mean that we would refactor the AppConfig to be provided from the context? Doesn't it mean that it would require a lot more refactoring than this? Could it be like a next step for this, or something?
> 
> I actually tried to make an interface for the AppConfig that could have been used with the Helper-classes, but JS/TS cannot have an interface-field for static methods, so I needed to drop that idea.

Not sure about the AppConfig case, but generally is it possible to replace static methods with singleton object instance, and then the class of the singleton object could have interface?

nenonja reviewed (approved):
Green enough for me! ✅  proxys are failing but known issue.
